### PR TITLE
UCP/KEEPALIVE: enabled by default

### PR DESCRIPTION
**Pull request by hoopoepg**
- enabled keepalive by default: 60s * 128 EP's
- added processing of UCP keepalive in iface_open call

### CODE REVIEW

**Comment by hoopoepg** [comment]
bot:pipe:retest

File: src/uct/tcp/tcp_iface.c [hunk 1]
```diff
@@ -580,10 +580,18 @@ static UCS_CLASS_INIT_FUNC(uct_tcp_iface_t, uct_md_h md, uct_worker_h worker,
...
     self->config.keepalive.intvl   = config->keepalive.intvl;
 
+    if ((config->keepalive.idle == UCS_MEMUNITS_AUTO) &&
+        (params->field_mask & UCT_IFACE_PARAM_FIELD_KEEPALIVE_INTERVAL)) {
+        self->config.keepalive.idle = params->keepalive_interval;
+    } else if (config->keepalive.idle == UCS_MEMUNITS_AUTO) {
+        self->config.keepalive.idle = ucs_time_from_sec(UCT_TCP_EP_DEFAULT_KEEPALIVE_IDLE);
+    } else {
+        self->config.keepalive.idle = config->keepalive.idle;
+    }
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/uct/tcp/tcp_iface.c:593]: ok, fixed

[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/tcp/tcp_iface.c:593]: looks same to me..

**Review by brminich — APPROVED** [review]

File: src/uct/tcp/tcp_iface.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/uct/tcp/tcp_iface.c:593]: hmmm, force-pushed again...

**Comment by hoopoepg** [comment]
resolved conflicts

File: src/ucp/core/ucp_context.c [hunk 2]
```diff
@@ -312,11 +312,11 @@ static ucs_config_field_t ucp_config_table[] = {
    ucs_offsetof(ucp_config_t, ctx.proto_enable), UCS_CONFIG_TYPE_BOOL},
 
   /* TODO: set for keepalive more reasonable values */
-  {"KEEPALIVE_INTERVAL", "0",
+  {"KEEPALIVE_INTERVAL", "60s",
```
Comment (yosefe) [inline src/ucp/core/ucp_context.c:315]: @hoopoepg will there be keepalive operation if all ep's are without error handling support, for example MPI case?

Comment (hoopoepg) [reply src/ucp/core/ucp_context.c:315]: ep_check will not be called, kernel based KA (TCP transport) will be enabled

Comment (yosefe) [reply src/ucp/core/ucp_context.c:315]: ok, i see that if all ep's are w/o ERR mode, KA (by ep_check) is not done in UCP

File: src/ucp/core/ucp_worker.inl [hunk 3]
```diff
@@ -122,6 +122,13 @@ ucp_worker_extract_request_by_id(ucp_worker_h worker, ucs_ptr_map_key_t id)
     return request;
 }
 
+static UCS_F_ALWAYS_INLINE int
+ucp_worker_keepalive_is_enabled(ucp_worker_h worker)
+{
+    return (worker->context->config.keepalive_interval != 0) &&
+           (worker->context->config.ext.keepalive_num_eps != 0);
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.inl:129]: per discussion with @brminich , need to check only worker->context->config.keepalive_interval != 0 here,
and if user sets worker->context->config.ext.keepalive_num_eps to 0, it's considered invalid value and need to fail

Comment (hoopoepg) [reply src/ucp/core/ucp_worker.inl:129]: ok, done

**Comment by hoopoepg** [comment]
@yosefe more comments?

**Review by dmitrygx — APPROVED** [review]

File: src/ucp/core/ucp_context.c [hunk 4]
```diff
@@ -1543,6 +1543,13 @@ static ucs_status_t ucp_fill_config(ucp_context_h context,
         }
     }
 
+    if ((context->config.ext.keepalive_interval != 0) &&
```
Comment (yosefe) [inline src/ucp/core/ucp_context.c:1546]: can check just keepalive_num_eps

File: src/ucp/core/ucp_worker.inl [hunk 5]
```diff
@@ -122,6 +122,16 @@ ucp_worker_extract_request_by_id(ucp_worker_h worker, ucs_ptr_map_key_t id)
     return request;
 }
 
+static UCS_F_ALWAYS_INLINE int
+ucp_worker_keepalive_is_enabled(ucp_worker_h worker)
+{
+    ucs_assert((worker->context->config.keepalive_interval == 0) ||
+               ((worker->context->config.keepalive_interval != 0) &&
+                (worker->context->config.ext.keepalive_num_eps != 0)));
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.inl:130]: better just assert worker->context->config.ext.keepalive_num_eps in the keepalive worker function

File: src/ucp/core/ucp_context.c (hunk 4 above)
Comment (hoopoepg) [reply src/ucp/core/ucp_context.c]: ok, updated

File: src/ucp/core/ucp_worker.inl (hunk 5 above)
Comment (hoopoepg) [reply src/ucp/core/ucp_worker.inl]: done

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
bot:pipe:retest

**Review by yosefe — APPROVED** [review]

**Pull request merged**